src/search/tasks/negated_axioms_task.cc

```diff
@@ -0,0 +1,358 @@
+        conditions_as_cnf.emplace_back();
+        for (FactProxy fact : axiom.get_effects()[0].get_conditions()) {
+            int var_id = fact.get_variable().get_id();
+            int num_vals = task_proxy.get_variables()[var_id].get_domain_size();
+            for (int value = 0; value < num_vals; ++value) {
+                if (value != fact.get_value()) {
+                    conditions_as_cnf.back().insert({var_id, value});
+                }
+            }
+        }
+    }
+
+    // We can see multiplying out the cnf as collecting all hitting sets.
+    set<FactPair> current;
+    set<set<FactPair>> hitting_sets;
+    collect_non_dominated_hitting_sets_recursively(
+        conditions_as_cnf, 0, current, hitting_sets);
+
+    for (const set<FactPair> &c : hitting_sets) {
+        negated_axioms.emplace_back(
+            head, vector<FactPair>(c.begin(), c.end()));
+    }
+}
+
+
+void NegatedAxiomsTask::collect_non_dominated_hitting_sets_recursively(
+    const vector<set<FactPair>> &conditions_as_cnf, size_t index,
+    set<FactPair> &hitting_set, set<set<FactPair>> &results) {
+    if (index == conditions_as_cnf.size()) {
+        /*
+           Check whether the axiom body denoted in body is dominated.
+           This is the case if there is a fact such that no conjunction
+           in the cnf is covered by *only* this fact, implying that
+           removing the fact from the body would still cover all
+           conjunctions.
+        */
+        set<FactPair> not_uniquely_used(hitting_set);
+        for (const set<FactPair> &condition : conditions_as_cnf) {
+            vector<FactPair> intersection;
+            set_intersection(condition.begin(), condition.end(),
+                             hitting_set.begin(), hitting_set.end(),
+                             back_inserter(intersection));
+            if (intersection.size() == 1) {
+                not_uniquely_used.erase(intersection[0]);
+            }
+        }
+        if (not_uniquely_used.empty()) {
+            results.insert(hitting_set);
+        }
+        return;
+    }
+
+    const set<FactPair> &conditions = conditions_as_cnf[index];
+    for (const FactPair &fact_pair : conditions) {
+        // The current set is covered with the current hitting set elements.
+        if (hitting_set.find(fact_pair) != hitting_set.end()) {
+            collect_non_dominated_hitting_sets_recursively(
+                conditions_as_cnf, index + 1, hitting_set, results);
+            return;
+        }
+    }
+
+    for (const FactPair &elem : conditions) {
+        // We don't allow choosing different facts from the same variable.
+        bool same_var_occurs_already = false;
+        for (const FactPair &f : hitting_set) {
+            if (f.var == elem.var) {
+                same_var_occurs_already = true;
+                break;
+            }
+        }
+        if (same_var_occurs_already) {
+            continue;
+        }
+
```

Comment:
I added the set `hitting_set_vars` and now just check if elem.var is in this set; is that what you meant?
At the same time I change `hitting_set` to also just add and remove elements instead of copying it all the time.